### PR TITLE
[Spec Decode] EAGLE 3: Micro-optimization of Draft to Target logit conversion 

### PR DESCRIPTION
**Pull request by ekagra-ranjan**
E3 needs to map the draft logit ids to target logit ids. Currently, the following operations happen:
1. `base = torch.arange(self.config.draft_vocab_size, device=logits.device)`
2. allocate new activation of dim [`bs`, `target_vocab_size`] in every fwd pass
3. fill the `target_vocab_size` with `draft_vocab_size` and remaning with -inf
4. argmax the tensor with `target_vocab_size` elements

This PR instead does:
1. argmax the tensor with `draft_vocab_size` elements (faster than step 4) since `draft_vocab_size` < `target_vocab_size`
2. find the offset in O(1) and adds the offset

It avoid extra allocation and computation for the  Draft to Target logit conversion in Eagle 3.

The PR also does some refactoring so that there is more code sharing bw model definitions. E.g., any model implementing E3 does not need to reimplement the Draft to Target logit conversion logic in the model definition.

### Test

Compute the AL and ensured it matches the AL w/o this PR.

E1:
`python3 examples/offline_inference/spec_decode.py --method eagle --num_spec_tokens 3 --dataset-name hf --dataset-path philschmid/mt-bench --num-prompts 80 --temp 0 --top-p 1.0 --top-k -1 --tp 1 --enable-chunked-prefill
`

output
```
--------------------------------------------------
total_num_output_tokens: 16956
num_drafts: 7405
num_draft_tokens: 22215
num_accepted_tokens: 9557
mean acceptance length: 2.29
--------------------------------------------------
```

E3
`python3 examples/offline_inference/spec_decode.py --method eagle3 --num_spec_tokens 3 --dataset-name hf --dataset-path philschmid/mt-bench --num-prompts 80 --temp 0 --top-p 1.0 --top-k -1 --tp 1 --enable-chunked-prefill
`

output
```

--------------------------------------------------                                                                                                             
total_num_output_tokens: 16930                                                                                                                                 
num_drafts: 6079                                                                                                                                               
num_draft_tokens: 18237                                                                                                                                        
num_accepted_tokens: 10840                                                                                                                                     
mean acceptance length: 2.78                                                                                                                                   
-------------------------------
```

### Benchmark
BS 64: 
* K=5: 17.82 -> 17.6
* K=7: 20.89 -> 20.62

1% improvement

### CODE REVIEW

File: vllm/v1/spec_decode/eagle_mixin.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import torch
+from typing import Optional
+
+class EagleMixin:
+    def propose_chain(self, hidden_states: torch.Tensor) -> torch.Tensor:
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle_mixin.py]: `propose_chain` feels like a poor name for this function, as it conflicts with `EagleProposer.propose()`. How about something closer to "get_draft_logits" or similar?

Comment (benchislett) [reply vllm/v1/spec_decode/eagle_mixin.py]: Or "greedy_sample_draft_tokens" might be more appropriate since I now see that this is a shortcut for compute_logits + greedy sampling

**Comment by benchislett** [comment]
does this optimization lead to any measurable speedup? Could you include benchmarking and/or profiling results as needed to illustrate the advantage?

File: vllm/v1/spec_decode/eagle_mixin.py (hunk 1 above)
Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle_mixin.py]: There is a method called [propose_tree](https://github.com/vllm-project/vllm/blob/61aedb5ffe056f83b1edab15610a644d32f40071/vllm/v1/spec_decode/eagle.py#L569) so propose_chain is analogious to it. The `propose_chain` and `propose_tree` are called from within `EagleProposer.propose()`. The purpose of this is to get the chain draft so I am not sure get_draft_logits is a better name

Comment (benchislett) [reply vllm/v1/spec_decode/eagle_mixin.py]: In what way is `propose_chain` here analogous to `propose_tree`? The latter seems like a much more comprehensive `propose()`-style function, whereas `propose_chain` here seems like a much simpler combination of `compute_logits` and greedy sampling?

Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle_mixin.py]: > In what way is propose_chain here analogous to propose_tree?

`propose_tree()` is called inside `propose()` and returns draft token ids at the current step
`propose_chain()` is called inside `propose()` and returns draft token ids at the current step

The functionality/role of `propose_chain()` is analogous to `propose_tree()`, i.e. its called by propose() and returns the draft tokens ids sampled at the current step. 

For E1, its compute logits and argmax whereas in E3 it has an extra step of offset computation. This different behavior of E1 and E3 is abstracted away from `propose()` function which will call `propose_chain()`.

Comment (benchislett) [reply vllm/v1/spec_decode/eagle_mixin.py]: I see what you mean. I think I would still prefer something like `sample_chain` or `propose_sample_greedy` or similar, but I am okay to go with `propose_chain` if you think it is the best choice.

File: vllm/v1/spec_decode/eagle_mixin.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            logits.shape[0],
+            self.config.vocab_size,
+        ), float('-inf'))
+        logits_new[:, targets] = logits
+        return logits_new
+
+    def propose_chain(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        logits = self.compute_draft_logits(hidden_states)
+        draft_token_ids = logits.argmax(dim=-1)
+        offset = self.draft_id_to_target_id[draft_token_ids]
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle_mixin.py]: pre-commit can be resolved here by adding `assert self.draft_id_to_target_id is not None`, since the check can't tell if it is allowed to be `None` here.

File: vllm/v1/spec_decode/eagle_mixin.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        targets = base + self.draft_id_to_target_id
+        logits_new = logits.new_full((
+            logits.shape[0],
+            self.config.vocab_size,
+        ), float('-inf'))
+        logits_new[:, targets] = logits
+        return logits_new
+
+    def propose_chain(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        logits = self.compute_draft_logits(hidden_states)
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle_mixin.py]: I think this is supposed to call `compute_logits`? Otherwise I don't see where `Eagle3Mixin.compute_logits` would ever get called here.
```suggestion
        logits = self.compute_logits(hidden_states)
```

**Comment by rahul-tuli** [comment]
Hi @ekagra-ranjan are you still working on this?

File: vllm/v1/spec_decode/eagle_mixin.py (hunk 1 above)
Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle_mixin.py]: Alright, I have renamed it to `sample_chain` following your suggestion.

File: vllm/v1/spec_decode/eagle_mixin.py (hunk 3 above)
Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle_mixin.py]: `compute_draft_logits()` is the optimized version of `compute_logits()` and is used in chain sampling. The original `compute_logits()` is still called by the Tree sampling [here](https://github.com/vllm-project/vllm/pull/25917/changes#diff-a4809a837fbf535a8f0999b11087a53ec1c53948b50c0a1fe64396bc86de9461R368) in `eagle.py`

File: vllm/v1/spec_decode/eagle_mixin.py (hunk 2 above)
Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle_mixin.py]: Thanks, fixed!

**Comment by ekagra-ranjan** [comment]
synced and resolved conflict

File: vllm/v1/spec_decode/eagle_mixin.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        return logits_new
+
+    def sample_chain(
+        self, hidden_states: torch.Tensor, use_local_argmax_reduction: bool = False
+    ) -> torch.Tensor:
+        if use_local_argmax_reduction and not (
+            hasattr(self, "draft_id_to_target_id")
+            and self.draft_id_to_target_id is not None
+        ):
+            return self.logits_processor.get_top_tokens(self.lm_head, hidden_states)
```
Comment (ekagra-ranjan) [inline vllm/v1/spec_decode/eagle_mixin.py:68]: this if block comes from recently merged: https://github.com/vllm-project/vllm/pull/34049 whereas the else part is what this PR originally added
